cosmoz-sse.js

```diff
@@ -1,230 +1,87 @@
-// @license Copyright (C) 2019 Neovici AB - Apache 2 License
-import { PolymerElement } from '@polymer/polymer/polymer-element';
-
-const SSEHandlers = {
-	text: data => data,
-	json: data => JSON.parse(data)
-};
-
-/**
-	`<cosmoz-sse>` is an element that allows subscribing to server-sent events.
-
-	@customElement
-	@polymer
-	@demo demo/index.html
-*/
-class CosmozSSE extends PolymerElement {
-	static get is() {
-		return 'cosmoz-sse';
-	}
-
-	static get properties() {
-		return {
-			_configuration: {
-				type: Object,
-				computed: '_computeConfiguration(withCredentials)'
-			},
-			/**
-			 * The SSE endpoint.
-			 */
-			url: {
-				type: String
-			},
-
-			/**
-			 * The named events you wish to subscribe to.
-			 */
-			events: {
-				type: Array,
-				value: [],
-				observer: '_onEventsChange'
+// @license Copyright (C) 2020 Neovici AB - Apache 2 License
+import {
+	component, useEffect, useMemo, useState
+} from 'haunted';
+
+const getJson = input => {
+		try {
+			return JSON.parse(input);
+		} catch (e) {
+			// eslint-disable-next-line no-console
+			console.error('cant parse JSON:', input);
+		}
+	},
+	SSEHandlers = {
+		text: data => data,
+		json: data => getJson(data)
+	},
+
+	// eslint-disable-next-line max-lines-per-function
+	CosmozSSE = function ({
+		events, withCredentials = false, handleAs = 'json', url
+	}) {
+		const fireEvent = (type, eventInit) => {
+				this.dispatchEvent(new CustomEvent(type, eventInit));
 			},
-
-			/**
-			 * How to parse the event data.
-			 * @type {('json'|'text')}
-			 */
-			handleAs: {
-				type: String,
-				value: 'json'
+			configuration = useMemo(() => ({
+				withCredentials
+			}), [withCredentials]),
+			[source, setSource] = useState(null),
+			eventTypes = useMemo(() => events && getJson(events) || [], [events]),
```

Comment:
Maybe we should drop the `events` entirely and just keep the `message` dispatch. That's how cz-frontend uses cosmoz-sse anyway...

Reply:
On second thought, named events are a part of the EventSource spec, so we should support these for third-party users.
My main gripe is with the way we receive the event list: as a stringified array. I think that is a relic from the tests for the polymer version, but polymer converted that to an actual array before passing it to the code. I think we should change this. We can do either:
1. a simple joined list
<cosmoz-sse url="xxx" events="ping,pong">
2. don't support setting events as an attribute, only as a property, which should not be a bother for lit or polymer usage:
lit`<cosmoz-sse url="xxx" .events="${array}" on-ping="${handler}">`
polymer`<cosmoz-sse url="xxx" events="[[array]]" on-ping="onPing">`

Reply:
Yeah I think the best would be your suggestion about light-dom children
```js
<cosmoz-sse url="..">
  <cosmoz-sse-subscription event="ping" on-event="onPing"></cosmoz-sse-subscription>
</cosmoz-sse>
```

Reply:
How will you implement this?

I have an idea of using an effect in the cosmoz-sse-subscription elements, so they register themselves on the parent node when connected and de-register when disconnected.
```js
useEffect(() => {
  if(!this.parentNode) {
    return
  }
  const onEvent = (ev) => fire(new CustomEvent(event, {...});
  this.parentNode.subscribe(event, onEvent);
  return () => this.parentNode.unsubscribe(event, onEvent);
}, [this.parentNode]);
```

Reply:
@megheaiulian what do you think about this?

Reply:
If you declare both elements in the same file you don't need to touch the DOM :smile: 

```js
useEffect(() => {
  if(!this.parentNode) {
    return;
  }
  registeredSubscribers.push(this);
  return () => registeredSubscribers.splice(thisIndex, 1);
}, [this.parentNode]);
```

Reply:
What about multiple cosmoz-sse elements in the same page?